src/Nethermind/Nethermind.BeaconNode.Peering/MothraPeeringWorker.cs

```diff
@@ -0,0 +1,264 @@
+                                {
+                                    return suffix;
+                                }
+
+                                return 0;
+                            })
+                            .DefaultIfEmpty()
+                            .Max();
+                        int newSuffix = existingSuffix + 1;
+                        string logDirectoryName = $"log{newSuffix:0000}";
+                        
+                        if (_logger.IsDebug()) LogDebug.CreatingMothraLogDirectory(_logger, logDirectoryName, baseDirectoryInfo.FullName, null);
+                        IDirectoryInfo logDirectory = baseDirectoryInfo.CreateSubdirectory(logDirectoryName);
+                        _logDirectoryPath = logDirectory.FullName;
+                    }
+                }
+            }
+
+            return _logDirectoryPath;
+        }
+
+        private async void HandleBeaconBlockAsync(SignedBeaconBlock signedBeaconBlock)
+        {
+            try
+            {
+                if (_mothraConfigurationOptions.CurrentValue.LogSignedBeaconBlockJson)
+                {
+                    string logDirectoryPath = GetLogDirectory();
+                    string fileName = string.Format("signedblock{0:0000}_{1}.json", (int) signedBeaconBlock.Message.Slot, signedBeaconBlock.Signature.ToString().Substring(0,10));
+                    string path = _fileSystem.Path.Combine(logDirectoryPath, fileName);
+                    using (Stream fileStream = _fileSystem.File.OpenWrite(path))
+                    {
+                        await JsonSerializer.SerializeAsync(fileStream, signedBeaconBlock, _jsonSerializerOptions).ConfigureAwait(false);
+                    }
+                }
+                
+                await _forkChoice.OnBlockAsync(_store, signedBeaconBlock).ConfigureAwait(false);
+            }
+            catch (Exception ex)
+            {
+                if (_logger.IsError())
+                    Log.HandleSignedBeaconBlockError(_logger, signedBeaconBlock.Message, ex.Message, ex);
+            }
+        }
+
+        private void OnGossipReceived(ReadOnlySpan<byte> topicUtf8, ReadOnlySpan<byte> data)
+        {
+            try
+            {
+                // TODO: handle other topics
+                if (topicUtf8.SequenceEqual(TopicUtf8.BeaconBlock))
+                {
+                    if (_logger.IsDebug())
+                        LogDebug.GossipReceived(_logger, nameof(TopicUtf8.BeaconBlock), data.Length, null);
+                    // Need to deserialize in synchronous context (can't pass Span async)
+                    SignedBeaconBlock signedBeaconBlock = Ssz.Ssz.DecodeSignedBeaconBlock(data);
+                    // TODO: maybe use a blocking queue that the receiver converts and writes to, then the async worker (ExcecuteAsync) can process.
+                    if (!ThreadPool.QueueUserWorkItem(HandleBeaconBlockAsync, signedBeaconBlock, true))
```

Comment:
any risk for queuing async block handling method on the thread pool? how will it handle exception?

Reply:
The work item, HandleBeaconBlock, like all event handlers, has nothing to pass exceptions up to (return is async void, not async Task), so does require a wrapping try/catch to log any exceptions (which it has).

The other option would be Task.Wait, to block the caller, but I don't think there is anything useful that it could do except either apart from catch and log any unhandled exception. 

Note that OnGossipReceived itself is an event handler, with the same issue, and is ultimately raised from the unmanaged callback from Mothra.